### PR TITLE
Chapter 14

### PR DESCRIPTION
**Pull request by VivianaSara**
### What is new in Code

-  In class **Restaurant** (from **Model**) I added 2 more attributes: **phone** and **description**
-  I update the initialization of restaurants variable (from **RestaurantTableViewController**)
-  I added 2 custom fonts
- now **RestaurantDetailViewController** I have just 2 Outlets (**headerView** and **tableView**)
- ### headerView
Screenshot 2022-04-27 at 18 11 42

> For this outlet I create a class **RestaurantDetailHeaderView** ( **View** group ). 
> It (**RestaurantDetailHeaderView** ) contain 2 labels (**nameLabel** and **typeLabel**), one image (**headerImageView**) and one button ( heart shape : **heartButton**) witch shows us if the specific restaurant is added to favourites

- ### tableView
Screenshot 2022-04-27 at 18 19 03

>  This table outlet contains 2 cell and for everyone I created a new view class: **RestaurantDetailTextCell** (for descriptionLabel) and **RestaurantDetailTwoColumnCell** (for address and phone details)

  ### 
 ### 
### UI elements
Screenshot 2022-04-27 at 17 56 18

### CODE REVIEW

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 1]
```diff
@@ -9,21 +9,25 @@ import UIKit
 
 class RestaurantDetailViewController: UIViewController {
 
-    @IBOutlet private var restaurantImageView: UIImageView!
-    @IBOutlet private var restaurantLocation: UILabel!
-    @IBOutlet private var restaurantType: UILabel!
-    @IBOutlet private var restaurantName: UILabel!
+    @IBOutlet var tableView: UITableView!
+    @IBOutlet var headerView: RestaurantDetailHeaderView!
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: Outlets should be private

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 2]
```diff
@@ -9,21 +9,25 @@ import UIKit
...
-        restaurantName.text = restaurant.getName()
-        restaurantType.text = restaurant.getType()
-        restaurantLocation.text = restaurant.getLocation()
+        // Configure header view
+        headerView.configure(nameLabel: restaurant.getName(), typeLabel: restaurant.getType(),
+                             headerImageView: restaurant.getImage())
+        let heartImage = restaurant.getIsFavorite() ? "heart.fill" : "heart"
+        headerView.setHeartButton(color: restaurant.getIsFavorite() ? .systemYellow : .white)
+        headerView.setHeartButton(image: heartImage)
+
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: Maybe you can pass the restaurant as a variable to the configure method and get everything you need from there in  RestaurantDetailHeaderView and set the heart button there too

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 3]
```diff
@@ -33,15 +37,41 @@ class RestaurantDetailViewController: UIViewController {
...
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 2
     }
-    */
 
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        switch indexPath.row {
+        case 0:
+            guard let cell = tableView.dequeueReusableCell(withIdentifier:
+                                                            String(describing: RestaurantDetailTextCell.self),
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: ideally the value of the param should be on the same line with the param name

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 4]
```diff
@@ -33,15 +37,41 @@ class RestaurantDetailViewController: UIViewController {
...
+                                                           for: indexPath) as? RestaurantDetailTextCell else {
+                return UITableViewCell()
+            }
+
+            cell.setDescriptionLabel(text: restaurant.getDescrption())
+            return cell
+
+        case 1:
+            guard let cell = tableView.dequeueReusableCell(withIdentifier:
+                                                            String(describing: RestaurantDetailTwoColumnCell.self),
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: same here as the previous comment

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 5]
```diff
@@ -33,15 +37,41 @@ class RestaurantDetailViewController: UIViewController {
...
+                                                           for: indexPath) as? RestaurantDetailTwoColumnCell else {
+                return UITableViewCell()
+            }
+
+            cell.setColumn1TitleLabel(text: "Address")
+            cell.setColumn1TextLabel(text: restaurant.getLocation())
+            cell.setColumn2TitleLabel(text: "Phone")
+            cell.setColumn2TextLabel(text: restaurant.getPhone())
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: I think you can also have all this under one method where you send the restaurant object

File: FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+class RestaurantDetailHeaderView: UIView {
+
+    @IBOutlet private var headerImageView: UIImageView!
+    @IBOutlet private var nameLabel: UILabel! {
+        didSet {
+            nameLabel.numberOfLines = 0
+            if let customFont = UIFont(name: "Nunito-Bold", size: 40.0) {
+                nameLabel.font = UIFontMetrics(forTextStyle: .title1).scaledFont(for: customFont)
+            }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift]: numberOfLines and font can also be set from storyboard

File: FoodPin/FoodPin/View/RestaurantDetailTextCell.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+import UIKit
+
+class RestaurantDetailTextCell: UITableViewCell {
+
+    @IBOutlet private var descriptionLabel: UILabel! {
+        didSet {
+            descriptionLabel.numberOfLines = 0
+        }
+    }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailTextCell.swift]: same here, can be set directly from storyboard

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    @IBOutlet private var column2TextLabel: UILabel! {
+        didSet {
+            column2TextLabel.numberOfLines = 0
+        }
+    }
+
+    func setColumn1TitleLabel(text: String) { self.column1TitleLabel.text = text }
+    func setColumn1TextLabel(text: String) { self.column1TextLabel.text = text }
+    func setColumn2TitleLabel(text: String) { self.column2TitleLabel.text = text }
+    func setColumn2TextLabel(text: String) { self.column2TextLabel.text = text }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: Number of lines can be set from storyboard and you can make text uppercased in the setColumn methods, where you initially set the text.
Also, the name of the labels are not very intuitive, what are those label actually holding, a title, description, name?

**Review by ralumarusca — CHANGES_REQUESTED** [review]

File: FoodPin/FoodPin/View/RestaurantDetailTextCell.swift [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+//
+
+import UIKit
+
+class RestaurantDetailTextCell: UITableViewCell {
+
+    @IBOutlet private var descriptionLabel: UILabel!
+
+    func setDescriptionLabel(text: String) { self.descriptionLabel.text = text }
+
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTextCell.swift]: empty line

**Review by ralumarusca — APPROVED** [review]

**Review by paulpop10 — APPROVED** [review]

**Pull request merged**